### ISSUE TITLE
Trouble using instance profile and/or task roles for S3 access

### ISSUE THREAD

**Issue by cbutcosk** [cantaloupe-project/cantaloupe#410]
First, thanks to all the folks that work on this project--cantaloupe's mix of features, API, and deployment options has really made it a great service in our (typically medium-sized) DH projects.

We run Cantaloupe in a docker container via an AWS ECS task. Source JP2s are stored on an S3 bucket and the container picks up access credentials via the task role. On the 4.0.X releases this configuration works fine, but on upgrading to 4.1.X requests that had returned `200` for objects in the S3 bucket now return `403 Forbidden`. 

#288 indicates this should be a valid way to get credentials but something is newly unhappy as of 4.1. Is this a configuration anyone else is successfully using in production? I may have something misconfigured in IAM or ECS, but perhaps it's not just me. :)

We're working around with a key / secret but in case anyone else hits a similar problem, this issue on the minio java SDK repo (https://github.com/minio/minio-java/issues/756) seems to indicate that instance profiles aren't a supported credential path? Though it looks like they have a merged PR that handles the full AWS credential chain (https://github.com/minio/minio-java/pull/1038)

**Comment by adolski** [comment]
Hi @cbutcosk,

Thanks for the feedback. Glad you are otherwise enjoying Cantaloupe.

It looks like those Minio tickets are for native Minio support for an instance profile credentials provider. S3Source doesn't rely on Minio for this, instead using an `AWSCredentialsProvider` from the AWS SDK, and providing the resulting credentials to the `MinioClient`.

There was some refactoring of AWS/S3-related code from 4.0 to 4.1:

#### 4.0
* [AWSCredentialsProvider initialization](https://github.com/cantaloupe-project/cantaloupe/blob/84bd4027553cb5c86ec61ba19475a4b9ed22d6a8/src/main/java/edu/illinois/library/cantaloupe/util/AWSClientBuilder.java#L121)
    * This relies on [DefaultAWSCredentialsProviderChain](https://github.com/aws/aws-sdk-java/blob/1.11.291/aws-java-sdk-core/src/main/java/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.java) in AWS SDK 1.11.291
#### 4.1
* [AWSCredentialsProvider initialization](https://github.com/cantaloupe-project/cantaloupe/blob/release/4.1/src/main/java/edu/illinois/library/cantaloupe/util/AWSClientBuilder.java#L53)
    * DefaultAWSCredentialsProviderChain is no longer used

I'm not sure what is happening in your case, but the first thing I would do is make sure that none of the providers added before the `EC2ContainerCredentialsProviderWrapper` in the last link above are providing credentials, as this would cause the `EC2ContainerCredentialsProviderWrapper`'s credentials to be overridden/ignored. I don't know why migrating from 4.0 to 4.1 would have broken that, though.

**Comment by cbutcosk** [comment]
@adolski Thanks for pointing me to the provider chain, it got me far enough to dig to the roots of this issue when I got a chance to look closer last week. 

Of course you're right that the provider chain is mostly unchanged--in 4.1.7 it correctly picks up the credentials from `EC2ContainerCredentialsProviderWrapper` and they match the keys I can pull off of container metadata manually. In looking at how MinioClient is handling things, I suspect the current dependency is not applying the appropriate session token header + signer workflow for [the temporary credentials](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_use-resources.html#RequestWithSTS) vended in [the IAM Role](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) auth flow. Basically MinioClient isn't signing in the right circumstances (in pre-8.0 it still assumes HTTPS reqs can be default unsigned and doesn't handle the additional header signing bits) and doesn't take a session token anyway. 

[In a very recent release](https://github.com/minio/minio-java/releases/tag/8.0.0) the Minio folks improved the knottier aspects of auth session management (session tokens and credential expiry though also STS assumed roles). As a proof of concept if I bump MinioClient to 8.0.0 and in `S3Source` coerce the `AWSCredentials` to an AWS SDK `BasicSessionCredentials` object I can get an appropriate session token and apply it with the MinioClient instance with their `StaticProvider` and things work as expected:
```java
    static synchronized MinioClient getClientInstance() {
        if (client == null) {
            final Configuration config = Configuration.getInstance();
            final AWSCredentialsProvider credentialsProvider = AWSClientBuilder.newCredentialsProvider(
                            config.getString(Key.S3SOURCE_ACCESS_KEY_ID),
                            config.getString(Key.S3SOURCE_SECRET_KEY));

            final AWSCredentials credentials =
                    credentialsProvider.getCredentials();

            final BasicSessionCredentials sessionCreds = (BasicSessionCredentials) credentials;

            LOGGER.debug("Got AWS session credentials key_id: {} session: {}",sessionCreds.getAWSAccessKeyId(),sessionCreds.getSessionToken());
            
            final String endpoint = config.getString(Key.S3SOURCE_ENDPOINT);
            final StaticProvider provider = new StaticProvider( sessionCreds.getAWSAccessKeyId(), 
                                                                sessionCreds.getAWSSecretKey(), 
                                                                sessionCreds.getSessionToken()
                                                                );

            client = MinioClient.builder()
                            .endpoint(endpoint)
                            .region(awsRegionFromURL(endpoint))
                            .credentialsProvider(provider)
                            .build();
            
        }
        return client;
    }
```

But obviously the unhandled complexity around session management is still unhandled here and that cast will only succeed on providers with temporary credentials. It's possible other-and-or-older `EC2ContainerCredentialsProviderWrapper` use cases don't expect a token--there is a code path in the wrapped `ContainerCredentialsProvider` that will return key/secret and no token. That said shimming the AWS SDK provider to handle that complexity and work as a Minio credentials provider looks quite doable.

**Comment by cbutcosk** [comment]
In the end we ended up fixing this with a subclass of the Minio `Provider` that passes their `fetch()` calls to the underlying AWS SDK `CredentialsProvider` class and checks if the credentials are a session-aware subclass of `BasicCredentials`. The SDK makes some guarantees about the credentials refresh being timeout-aware for credential types that have timeouts (and potential IO dependencies). [It kept us from messing with too much else](https://gist.github.com/cbutcosk/d0b706cb381666404c4907c46b39a7ca) in the codebase--I'm cautiously optimistic thus far.

**Comment by adolski** [comment]
Thanks for looking into this, @cbutcosk. If this is working out for you, I'd be happy to accept a PR for inclusion in 4.1.8.

On the `develop` branch, I've removed both Minio and AWS SDK 1.0 (used by S3Cache) and replaced them with AWS SDK 2.0, so this shouldn't be a problem there. (There is some backstory to why Minio is being used in 4.1.x, but it's probably off-topic.)

**Comment by adolski** [comment]
It turns out that version 5.0 contains a bug that prevents this from working, but it should be fixed on `release/5.0` for inclusion in 5.0.1.

I don't really have time to deal with this in the 4.1 series, but I will still accept a pull request.
